### DISCUSSION TITLE
Why do tests with no asserts or expects pass?

### DISCUSSION THREAD

**Discussion by designermonkey** [vitest-dev/vitest#5361] [General]
I've been doing unit testing for years in PHP and Java and if you define a test in those languages without asserting that something happened, the test fails. This is what I thought all unit testing frameworks did. Then I come to `vitest`.

Can someone explain to me why, if I have a `test` in my file with nothing in it, it passes?

```javascript
import { test } from 'vitest'

test('surely this should fail?', () => {

})
```

**Comment by sheremet-va** [comment]
The test fails only if the function throws an error. An empty function doesn't throw an error. With this approach you can use any assertion library you like, or just throw an error manually if needed.

**Reply by designermonkey** [reply]
This seems wrong. It allows a team to list a load of tests that describe functionality, write no actual test code, and maintain a passing test suite.

Unit test are supposed to fail if you don't assert something, so if expect/assert is never called inside a test, it should fail surely.

**Reply by designermonkey** [reply]
TDD challenges us to follow the principle of fail, pass, refactor, but we can't do that if we are presented with pass, pass, refactor.

**Reply by sheremet-va** [reply]
As I said, the framework (Vitest) cannot track assertions because you can use anything as an assertion. We provide `expect.hasAssertions()` API that you can call if you only use `expect`.

**Reply by nmay231** [reply]
@sheremet-va Then can we enable a flag that applies `expect.hasAssertions()` for all tests, if you only use `expect`?

If we need to add an opt-out, we could use `expect.assertions(0)` or `expect.assertions("unknown")` if we don't know how many assertions there could be.

**Reply by sheremet-va** [reply]
I guess we can add an option for this. For now, you can call it inside beforeEach for the same effect

**Reply by sheremet-va** [reply]
Should be addressed in #5729
